sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -168,6 +170,85 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]] extends TreeNode[PlanT
     }.toSeq
   }
 
+
+  /**
+   * Rewrites this plan tree based on the given plan mappings from old plan nodes to new nodes.
+   * This method also updates all the related references in this plan tree accordingly, in case
+   * the replaced node has different output expr ID than the old node.
+   */
+  def rewriteWithPlanMapping(
+      planMapping: Map[PlanType, PlanType],
+      canGetOutput: PlanType => Boolean = _ => true): PlanType = {
+    def internalRewrite(plan: PlanType): (PlanType, Seq[(Attribute, Attribute)]) = {
+      if (planMapping.contains(plan)) {
```

Comment:
IIUC this check cannot correctly handle nested cases in `planMapping`; for example,
```
Project
 +- Union
    :+- (1) Project 
    :   +- Union
    :   :  :+- (2) Project
    :   :
    :   +- Project
    :
    +- Project
       +- ...
```
If the two nested `Project`s above, `(1)` and `(2)`,  are stored in `planMapping`, I think only the case `(1)` is matched in this condition then the case `(2)` is just ignored. So, I rewrote the logic a bit so that plans are replaced in a bottom-up way in the previous PR:
https://github.com/apache/spark/blob/1de272f98d0ff22d0dd151797f22b8faf310963a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L140-L144

Reply:
hmm, is this a real-world case? I think this is too complicated if we need to replace nodes in the values of `planMapping`.

Reply:
hm, yea, this is a little complicated though, I remember the existing tests fail because of this reason. That might be the case below;
```
SQLQueryTestSuite.sql
org.scalatest.exceptions.TestFailedException: union.sql
Expected "struct<[c1:decimal(11,1),c2:string]>", but got "struct<[]>" Schema did not match for query #3
SELECT *
FROM   (SELECT * FROM t1
        UNION ALL
        SELECT * FROM t2
        UNION ALL
        SELECT * FROM t2): -- !query
SELECT *
FROM   (SELECT * FROM t1
        UNION ALL
        SELECT * FROM t2
        UNION ALL
        SELECT * FROM t2)
-- !query schema
struct<>
-- !query output
org.apache.spark.sql.catalyst.errors.package$TreeNodeException
After applying rule org.apache.spark.sql.catalyst.optimizer.RemoveNoopOperators in batch Operator Optimization before Inferring Filters, the structural integrity of the plan is broken., tree:
'Union false, false
```

Reply:
you are right, this is a valid use case.